matrix_factorization/loss.cu

```diff
@@ -9,30 +9,35 @@
 #include "matrix.h"
 
 #define index(i, j, N)  ((i)*(N)) + (j)
+#define warp_size 32 //TODO: we need to get device props
 
 using namespace cu2rec;
 
 // PARALLEL
+extern __shared__ float biases[];
 __global__ void loss_kernel(int factors, int user_count, int item_count, const float * P, const float * Q, const int * indptr, 
                             const int * indices, const float * data, float * error, float * user_bias, float * item_bias, float global_bias) {
+
+    // use first warp to load in user_biases
+    float* s_user_bias = (float*)biases;
+    if(threadIdx.x < warp_size) {
+        for(int i = 0; i < user_count; i += warp_size) {
```

Comment:
So, this doesn't use anything related to current thread at all, meaning this gets repeated for all threads. Which also makes me think that only the first user of each warp is loaded into the shared memory. Instead, maybe we can let all threads load in their own biases to shared mem?